src/value_serializer.rs

```diff
@@ -0,0 +1,374 @@
+    let embedder = this.offset_within_embedder.to_embedder_mut::<Opaque>(this);
+    std::mem::transmute((embedder, vtable))
+  }
+
+  pub unsafe fn dispatch_box(
+    value_serializer_delegate: &mut ValueSerializerDelegate,
+  ) -> Box<dyn ValueSerializerDelegateImpl> {
+    std::mem::transmute(Self::dispatch_mut(value_serializer_delegate))
+  }
+}
+
+// ValueSerializer
+
+// Must be == sizeof(v8::ValueSerializer), see v8__ValueSerializer__CONSTRUCT().
+#[repr(C)]
+pub struct ValueSerializer([usize; 1]);
+
+extern "C" {
+  fn v8__ValueSerializer__CONSTRUCT(
+    buf: *mut MaybeUninit<ValueSerializer>,
+    isolate: *mut Isolate,
+  );
+  fn v8__ValueSerializer__CONSTRUCT_with_delegate(
+    buf: *mut MaybeUninit<ValueSerializer>,
+    isolate: *mut Isolate,
+    delegate: *mut ValueSerializerDelegate,
+  );
+
+  fn v8__ValueSerializer__DESTRUCT(this: *mut ValueSerializer);
+
+  fn v8__ValueSerializer__Release(
+    this: *mut ValueSerializer,
+    ptr: *mut *mut u8,
+    size: *mut usize,
+  );
+
+  fn v8__ValueSerializer__TransferArrayBuffer(
+    this: *mut ValueSerializer,
+    transfer_id: u32,
+    array_buffer: Local<ArrayBuffer>,
+  );
+
+  fn v8__ValueSerializer__WriteHeader(this: *mut ValueSerializer);
+  fn v8__ValueSerializer__WriteValue(
+    this: *mut ValueSerializer,
+    context: Local<Context>,
+    value: Local<Value>,
+  ) -> MaybeBool;
+  fn v8__ValueSerializer__WriteUint32(this: *mut ValueSerializer, value: u32);
+  fn v8__ValueSerializer__WriteUint64(this: *mut ValueSerializer, value: u64);
+  fn v8__ValueSerializer__WriteDouble(this: *mut ValueSerializer, value: f64);
+  fn v8__ValueSerializer__WriteRawBytes(
+    this: *mut ValueSerializer,
+    source: *const c_void,
+    length: usize,
+  );
+}
+
+impl ValueSerializer {
+  pub fn new(scope: &mut HandleScope<'_>) -> Self {
+    let isolate: *mut Isolate = scope.get_isolate_ptr();
+    let mut value_serializer: MaybeUninit<Self> = MaybeUninit::uninit();
+    unsafe {
+      v8__ValueSerializer__CONSTRUCT(&mut value_serializer, isolate);
+      value_serializer.assume_init()
+    }
+  }
+
+  pub fn new_with_delegate(
+    scope: &mut HandleScope<'_>,
+    delegate: &mut ValueSerializerDelegate,
+  ) -> Self {
+    let isolate: *mut Isolate = scope.get_isolate_ptr();
+    let mut value_serializer: MaybeUninit<Self> = MaybeUninit::uninit();
+    unsafe {
+      v8__ValueSerializer__CONSTRUCT_with_delegate(
+        &mut value_serializer,
+        isolate,
+        delegate,
+      );
+      value_serializer.assume_init()
+    }
+  }
+}
+
+impl Drop for ValueSerializer {
+  fn drop(&mut self) {
+    unsafe { v8__ValueSerializer__DESTRUCT(self) };
+  }
+}
+
+impl ValueSerializer {
+  pub fn transfer_array_buffer(
+    &mut self,
+    transfer_id: u32,
+    array_buffer: Local<ArrayBuffer>,
+  ) {
+    unsafe {
+      v8__ValueSerializer__TransferArrayBuffer(self, transfer_id, array_buffer)
+    };
+  }
+
+  pub fn release(&mut self) -> &[u8] {
```

Comment:
This function unfortunately leaks memory. You can go in two directions with this:

1. The ValueSerializer remains the owner of the buffer even after `release()` is called. It frees the buffer in it's drop handler.
2. The ValueSerializer is consumed by this function, and returns a buffer that is owned by the caller. What return type is appropriate depends on how the buffer is allocated.

Reply:
I think we have to go for option 2; as the v8 ValueSerializer won't be usable after calling this function. (https://github.com/v8/v8/blob/master/include/v8.h#L2373)

Reply:
> I think we have to go for option 2; as the v8 ValueSerializer won't be usable after calling this function. (https://github.com/v8/v8/blob/master/include/v8.h#L2373)

I think option 2 is indeed the more elegant choice. The unsolved question however is which "owned" type we should return. It would depend on the concrete implementation that the user provides for ReallocateBufferMemory/FreeBufferMemory. I see another complication too - the `ReallocateBufferMemory` method does not receive the "old" length of the buffer, so it might be difficult to provide an implementation that uses an appropriate rust type (`Vec<u8>` or `Box<[u8]>` would be the most likely candidates). 

I'll need to think about this a bit longer.

Reply:
I thought the `size` parameter given to the `ReallocateBufferMemory` function indicates the length of the old buffer?  
In the current implementation I mapped the `ReallocateBufferMemory` function to a rust function with the following interface:

```rust
fn reallocate_buffer_memory(
    &mut self,
    old_buffer: Option<&[u8]>,
  ) -> Option<&mut [u8]>;
```

So I'm currently using (optional) slices. I think `Vec<u8>` won't work because a vector also keeps track of the "used" memory space additional to the allocated space.
Maybe `Box<[u8]>` should be used instead of `&mut [u8]`? From what I can find in the rust docs, passing a box actually means transferring ownership, whereas a reference means your lending the memory.
I created a simple rust playground to better explain how we can convert pointer + size to a boxed slice: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5dfa5af617124fe545a45af3d46c8770

Reply:
> Maybe `Box<[u8]>` should be used instead of `&mut [u8]`? From what I can find in the rust docs, passing a box actually means transferring ownership,

Correct! `Box<[u8]>` would actually be great.
Unfortunately that, too, is a fat pointer.

Reply:
I changed the array slices to boxed slices.